### PR TITLE
[SHELL32] Invoke the ItemIDList of shortcuts

### PR DESCRIPTION
**Pull request by whindsaks**
Invoking the IDList as if the target was double-clicked in Explorer fixes shortcuts to control panel items (both .cpl and ShellFolder types).

Notes:
 - The shortcut has to be created by #6999 for shortcuts to .cpl files to work correctly.

JIRA issue: [CORE-19690](https://jira.reactos.org/browse/CORE-19690)

### CODE REVIEW

File: dll/win32/shell32/CShellLink.cpp [hunk 1]
```diff
@@ -2640,12 +2640,9 @@ HRESULT CShellLink::DoOpen(LPCMINVOKECOMMANDINFO lpici)
     sei.lpParameters = args;
     sei.lpClass = m_sLinkPath;
     sei.nShow = m_Header.nShowCommand;
+    if (lpici->nShow != SW_SHOWNORMAL && lpici->nShow != SW_SHOW)
+        sei.nShow = lpici->nShow; // Allow invoker to override .lnk show mode
```
Comment (HBelusca) [inline dll/win32/shell32/CShellLink.cpp:2644]: This is compatible with Windows behaviour?

File: dll/win32/shell32/CShellLink.cpp [hunk 2]
```diff
@@ -2640,12 +2640,9 @@ HRESULT CShellLink::DoOpen(LPCMINVOKECOMMANDINFO lpici)
...
     sei.lpDirectory = m_sWorkDir;
-    sei.lpVerb = L"open";
-
-    // HACK for ShellExecuteExW
-    if (m_sPath && wcsstr(m_sPath, L".cpl"))
-        sei.lpVerb = L"cplopen";
```
Comment (HBelusca) [inline dll/win32/shell32/CShellLink.cpp:2648]: If I understand correctly, this is now handled via the `SEE_MASK_IDLIST` ---> `SEE_MASK_INVOKEIDLIST` change (the crux of this PR)?

File: dll/win32/shell32/wine/classes.c [hunk 3]
```diff
@@ -130,9 +130,11 @@ BOOL HCR_MapTypeToValueA(LPCSTR szExtension, LPSTR szFileType, LONG len, BOOL bP
 	return TRUE;
 }
 
+EXTERN_C HRESULT SHELL32_EnumDefaultVerbList(LPCWSTR List, UINT Index, LPWSTR Verb, SIZE_T cchMax);
```
Comment (HBelusca) [inline dll/win32/shell32/wine/classes.c:133]: It looks like all your changes in this file need to be inside a `#ifdef __REACTOS__ ... #endif` block.

File: dll/win32/shell32/shlexec.cpp [hunk 4]
```diff
@@ -1553,20 +1553,31 @@ static HRESULT ShellExecute_ContextMenuVerb(LPSHELLEXECUTEINFOW sei)
...
         return hr;
 
+    BOOL fDefault = !sei->lpVerb || !sei->lpVerb[0];
+
     CComHeapPtr<char> verb, parameters;
     __SHCloneStrWtoA(&verb, sei->lpVerb);
     __SHCloneStrWtoA(&parameters, sei->lpParameters);
 
     CMINVOKECOMMANDINFOEX ici = {};
     ici.cbSize = sizeof ici;
```
Comment (HBelusca) [inline dll/win32/shell32/shlexec.cpp:1563]:
```suggestion
    ici.cbSize = sizeof(ici);
```

File: dll/win32/shell32/shlexec.cpp [hunk 5]
```diff
@@ -1553,20 +1553,31 @@ static HRESULT ShellExecute_ContextMenuVerb(LPSHELLEXECUTEINFOW sei)
     if (FAILED_UNEXPECTEDLY(hr))
         return hr;
 
+    BOOL fDefault = !sei->lpVerb || !sei->lpVerb[0];
```
Comment (HBelusca) [inline dll/win32/shell32/shlexec.cpp:1556]: Could you move this one a little bit further down, just before the `if (!fDefault)` ?

File: dll/win32/shell32/shlexec.cpp [hunk 6]
```diff
@@ -1553,20 +1553,31 @@ static HRESULT ShellExecute_ContextMenuVerb(LPSHELLEXECUTEINFOW sei)
...
+
     CComHeapPtr<char> verb, parameters;
     __SHCloneStrWtoA(&verb, sei->lpVerb);
     __SHCloneStrWtoA(&parameters, sei->lpParameters);
 
     CMINVOKECOMMANDINFOEX ici = {};
     ici.cbSize = sizeof ici;
-    ici.fMask = (sei->fMask & (SEE_MASK_NO_CONSOLE | SEE_MASK_ASYNCOK | SEE_MASK_FLAG_NO_UI));
+    ici.fMask = (sei->fMask & (SEE_MASK_NO_CONSOLE | SEE_MASK_ASYNCOK | SEE_MASK_FLAG_NO_UI)) | CMIC_MASK_UNICODE;
```
Comment (HBelusca) [inline dll/win32/shell32/shlexec.cpp:1564]: Does enabling unicode support as done there fix some extra stuff?

File: dll/win32/shell32/CShellLink.cpp (hunk 1 above)
Comment (whindsaks) [reply dll/win32/shell32/CShellLink.cpp:2644]: Yes

File: dll/win32/shell32/CShellLink.cpp (hunk 2 above)
Comment (whindsaks) [reply dll/win32/shell32/CShellLink.cpp:2648]: Yes, SEE_MASK_INVOKEIDLIST means use IContextMenu and this becomes the same as double-clicking in the folder.

File: dll/win32/shell32/shlexec.cpp (hunk 5 above)
Comment (whindsaks) [reply dll/win32/shell32/shlexec.cpp:1556]: No, this fixes a bug where ici.lpVerb was tested but it is sei->lpVerb that needs to be tested (`__SHCloneStrWtoA(&verb, sei->lpVerb);` hides the true NULL when that verb buffer is assigned to ici.lpVerb)

File: dll/win32/shell32/shlexec.cpp (hunk 6 above)
Comment (whindsaks) [reply dll/win32/shell32/shlexec.cpp:1564]: Setting that flag means the caller should check lpParametersW instead and this is of course required if the parameter string contains something that cannot be converted to CP_ACP.

Comment (HBelusca) [reply dll/win32/shell32/shlexec.cpp:1564]: ok yes I could imagine something like shortcut targets or program parameters specifying paths with non-latin-character names.

File: dll/win32/shell32/shlexec.cpp (hunk 5 above)
Comment (HBelusca) [reply dll/win32/shell32/shlexec.cpp:1556]: Ah, but I wasn't suggesting you use ici.lpVerb in the comparison. I was just suggesting moving this line verbatim just before the if (!fDefault).
The sei->lpVerb (pointer and its contents) aren't changes there, are they?

Comment (whindsaks) [reply dll/win32/shell32/shlexec.cpp:1556]: No but someone might be tempted to use ini.lpVerb instead (again) since it looks like a valid thing to do.

Comment (HBelusca) [reply dll/win32/shell32/shlexec.cpp]: understandable

**Review by HBelusca — APPROVED** [review]
LGTM; I would still suggest to put the ROS-specific code that is in the wine/ file inside `__REACTOS__` ifdefs, but I let other reviewers decide about this.

**Review by katahiromz — APPROVED** [review]
No objection from me.
